File: 2021/java/Day0106.java
/**
 * Author: Loading BG#7962
 */

import java.util.Arrays;

class Day0106 {
    public static void main(String[] args) {
        System.out.println("averageChar #1: " + averageChar("Hello"));
        System.out.println("averageChar #2: " + averageChar("Yes"));
        System.out.println("averageChar #3: " + averageChar("aaaa"));
        System.out.println("-------------");
        System.out.println("space #1: " + space("HelloWorld!"));
        System.out.println("space #2: " + space("HowAreYou?"));
        System.out.println("space #3: " + space("Example"));
        System.out.println("space #4: " + space("Simple,Right?"));
        System.out.println("-------------");
        System.out.println("sort #1: " + sort("Hello Yes aaaa"));
    }
    
    private static char averageChar(String s) {
        return (char) s.chars().average().getAsDouble();
    }
    
    private static String space(String sentence) {
        return sentence.charAt(0) + sentence.substring(1).replaceAll("([A-Z])", " $1").toLowerCase();
    }
    
    private static String sort(String sentence) {
        String[] words = sentence.split("\\s+");
        return Arrays.stream(words)
            .sorted((s1, s2) -> Double.compare(s1.chars().average().getAsDouble(), s2.chars().average().getAsDouble()))
            .collect(StringBuilder::new, (acc, s) -> acc.append(s).append(" "), StringBuilder::append).toString().trim();
    }
}

File: 2021/haskell/Day0106.hs
{-
  Author: Loading BG#7962
-}

module Main where

import Data.Char
import Data.List (sortBy)

avg :: [Int] -> Int
avg xs = div (foldl (+) 0 xs) (length xs)
averageChar :: String -> Char
averageChar = chr . avg . map ord

spaceHelper :: Int -> String -> String
spaceHelper i s
  | length s == 0    = ""
  | i == 0           = (s !! 0) : (spaceHelper (i + 1) $ drop 1 s)
  | isUpper (s !! 0) = " " ++ [toLower (s !! 0)] ++ (spaceHelper (i + 1) $ drop 1 s)
  | otherwise        = (s !! 0) : (spaceHelper (i + 1) $ drop 1 s)
space :: String -> String
space s = spaceHelper 0 s

sort :: String -> String
sort = unwords . sortBy (\a b -> compare (averageChar a) (averageChar b)) . words

main :: IO ()
main = do
  putStrLn ("averageChar #1: " ++ (show $ averageChar "Hello"))
  putStrLn ("averageChar #2: " ++ (show $ averageChar "Yes"))
  putStrLn ("averageChar #3: " ++ (show $ averageChar "aaaa"))
  putStrLn "-----------"
  putStrLn ("space #1: " ++ (show $ space "HelloWorld!"))
  putStrLn ("space #2: " ++ (show $ space "HowAreYou?"))
  putStrLn ("space #3: " ++ (show $ space "Example"))
  putStrLn ("space #4: " ++ (show $ space "Simple,Right?"))
  putStrLn "-----------"
  putStrLn ("sort #1: " ++ (show $ sort "Hello Yes aaaa"))
